File: 06-1-CheckWeatherApp_kt2/app/src/main/java/com/example/retrofit1app_kt2/MainViewModel.kt
package com.example.retrofit1app_kt2

import androidx.compose.runtime.mutableStateOf
import androidx.lifecycle.ViewModel
import androidx.compose.runtime.getValue
import androidx.compose.runtime.setValue
import androidx.lifecycle.viewModelScope
import com.example.retrofit1app_kt2.service.WeatherRepository
import kotlinx.coroutines.launch

class MainViewModel : ViewModel() {
    private val repository = WeatherRepository()

    var wheatherLabel by mutableStateOf("")
    var loading by mutableStateOf(false)

    fun getWeather(city: String) {
        val cityName = if (city.trim().isEmpty()) "São Carlos" else city
        wheatherLabel = ""
        loading = true

        viewModelScope.launch {
            try {
                val response = repository.getWeatherByCityName(cityName)
                val label = "Temperature: %.2f C"
                wheatherLabel = String.format(label, response.main.temp)
            } catch (e: Exception) {
                wheatherLabel = "Error: " + e.message
            } finally {
                loading = false
            }
        }
    }
}

File: flutter_projects/06-1-checkweather_fl/lib/main_view_model.dart
import 'package:checkweather_fl/service/weather_repository.dart';
import 'package:flutter/material.dart';

class MainViewModel extends ChangeNotifier {
  final WeatherRepository repository;

  MainViewModel({ required this.repository });

  // private mutable vars
  String _weatherLabel = "";
  String _weatherCity = "";
  bool _loading = false;

  // getters
  get weatherLabel => _weatherLabel;
  get loading => _loading;

  Future<void> getWeather(String city) async {
    String cityName = city.trim().isEmpty ? "São Carlos" : city;
    _weatherLabel = "";
    _loading = true;
    notifyListeners();

    try {
      var weatherResponse = await repository.getWeatherByCityName(cityName);
      double temp = weatherResponse.temp;
      _weatherCity = cityName;
      _weatherLabel = "Temperature: ${temp.toStringAsFixed(1)} °C";
    } catch (error) {
      _weatherLabel = "Error: $error";
    } finally {
      _loading = false;
      notifyListeners();
    }
  }

  void reset(String currentCityName) {
    if (_weatherLabel.isNotEmpty && currentCityName != _weatherCity) {
      _weatherLabel = "";
      notifyListeners();
    }
  }
}